Algorithms.Tests/Numeric/Decomposition/SVDTests.cs

```diff
@@ -0,0 +1,135 @@
+            Assert.AreEqual(1, Math.Abs(Svd.RandomUnitVector(1)[0]), epsilon);
+            Assert.AreNotEqual(Svd.RandomUnitVector(10), Svd.RandomUnitVector(10));
+        }
+
+        [Test]
+        public void Svd_Decompose()
+        {
+            CheckSvd(new double[,] { { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 } });
+            CheckSvd(new double[,] { { 1, 2, 3 }, { 4, 5, 6 } });
+            CheckSvd(new double[,] { { 1, 0, 0, 0, 2 }, { 0, 3, 0, 0, 0 }, { 0, 0, 0, 0, 0 }, { 0, 2, 0, 0, 0 } });
+        }
+
+        public double[,] GenerateRandomMatrix(int m, int n)
+        {
+            double[,] result = new double[m, n];
+            Random random = new Random();
+            for (int i = 0; i < m; i++)
+            {
+                for (int j = 0; j < n; j++)
+                {
+                    result[i, j] = random.NextDouble() - 0.5;
+                }
+            }
+            return result;
+        }
+
+        [Test]
+        public void Svd_Random([Random(3, 10, 5)] int m, [Random(3, 10, 5)] int n)
+        {
+            double[,] matrix = GenerateRandomMatrix(m, n);
+            CheckSvd(matrix);
+        }
+
+        public void CheckSvd(double[,] testMatrix)
+        {
+            double epsilon = 1E-5;
+            double[,] u;
+            double[,] v;
+            double[] s;
+            (u, s, v) = Svd.Decompose(testMatrix, 1E-8, 1000);
+
+            for (int i = 0; i < u.GetLength(1); i++)
+            {
+                double[] extracted = new double[u.GetLength(0)];
+                for (int j = 0; j < extracted.Length; j++)
+                {
+                    extracted[j] = u[j, i];
+                }
+
+                if (s[i] > epsilon)
+                {
+                    Assert.AreEqual(1, V.Magnitude(extracted), epsilon);
+                }
+                else
+                {
+                    Assert.Less(V.Magnitude(extracted), 1);
+                }
+            }
+
+            for (int i = 0; i < v.GetLength(1); i++)
+            {
+                double[] extracted = new double[v.GetLength(0)];
+                for (int j = 0; j < extracted.Length; j++)
+                {
+                    extracted[j] = v[j, i];
+                }
+
+                if (s[i] > epsilon)
+                {
+                    Assert.AreEqual(1, V.Magnitude(extracted), epsilon);
+                }
+                else
+                {
+                    Assert.Less(V.Magnitude(extracted), 1);
+                }
+            }
+
+            double[,] expanded = new double[s.Length, s.Length];
+            for (int i = 0; i < s.Length; i++)
+            {
+                expanded[i, i] = s[i];
+            }
+
+            AssertMatrixEqual(testMatrix,
+                M.MultiplyGeneral(M.MultiplyGeneral(u, expanded), M.Transpose(v)), epsilon);
+
+            AssertMatrixEqual(testMatrix,
+                M.MultiplyGeneral(u, M.MultiplyGeneral(expanded, M.Transpose(v))), epsilon);
```

Comment:
Could you also verify that s[0] >= s[1] >= s[2] >= ... >= s[last]?

Reply:
Sure